core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java

```diff
@@ -447,104 +415,144 @@ public AccessLogWriter accessLogWriter() {
+        for (;;) {
+            final HttpHeaders oldValue = additionalResponseHeaders;
+            final HttpHeaders newValue = oldValue.toBuilder().addObject(name, value).build();
+            if (additionalResponseHeadersUpdater.compareAndSet(this, oldValue, newValue)) {
+                return;
+            }
+        }
     }
 
     @Override
     public void addAdditionalResponseHeaders(Iterable<? extends Entry<? extends CharSequence, ?>> headers) {
         requireNonNull(headers, "headers");
-        additionalResponseHeaders = createAdditionalHeadersIfAbsent().toBuilder().addObject(headers).build();
+        for (;;) {
+            final HttpHeaders oldValue = additionalResponseHeaders;
+            final HttpHeaders newValue = oldValue.toBuilder().addObject(headers).build();
+            if (additionalResponseHeadersUpdater.compareAndSet(this, oldValue, newValue)) {
+                return;
+            }
+        }
     }
 
     @Override
     public boolean removeAdditionalResponseHeader(CharSequence name) {
         requireNonNull(name, "name");
-        final HttpHeaders additionalResponseHeaders = this.additionalResponseHeaders;
-        if (additionalResponseHeaders == null || additionalResponseHeaders.isEmpty()) {
-            return false;
-        }
+        for (;;) {
+            final HttpHeaders oldValue = additionalResponseHeaders;
+            if (oldValue.isEmpty() || oldValue.contains(name)) {
+                return false;
+            }
 
-        final HttpHeadersBuilder builder = createAdditionalHeadersIfAbsent().toBuilder();
-        final boolean removed = builder.remove(name);
-        this.additionalResponseHeaders = builder.build();
-        return removed;
+            final HttpHeaders newValue = oldValue.toBuilder().removeAndThen(name).build();
+            if (additionalResponseHeadersUpdater.compareAndSet(this, oldValue, newValue)) {
+                return true;
+            }
+        }
     }
 
     @Override
     public HttpHeaders additionalResponseTrailers() {
-        final HttpHeaders additionalResponseTrailers = this.additionalResponseTrailers;
-        if (additionalResponseTrailers == null) {
-            return HttpHeaders.of();
-        }
         return additionalResponseTrailers;
     }
 
     @Override
     public void setAdditionalResponseTrailer(CharSequence name, Object value) {
         requireNonNull(name, "name");
         requireNonNull(value, "value");
-        additionalResponseTrailers = createAdditionalTrailersIfAbsent().toBuilder()
-                                                                       .setObject(name, value).build();
+        for (;;) {
+            final HttpHeaders oldValue = additionalResponseTrailers;
+            final HttpHeaders newValue = oldValue.toBuilder().setObject(name, value).build();
+            if (additionalResponseTrailersUpdater.compareAndSet(this, oldValue, newValue)) {
+                return;
+            }
+        }
```

Comment:
nit: How about extracting this into private method?
```java
public void setAdditionalResponseTrailer(CharSequence name, Object value) {
    requireNonNull(name, "name");
    requireNonNull(value, "value");

    updateAdditionalResponseTrailer(oldValue -> oldValue.toBuilder().setObject(name, value).build());
}
private void updateAdditionalResponseTrailer(Function<HttpHeaders, HttpHeaders> updater) {
    for (;;) {
        final HttpHeaders oldValue = additionalResponseTrailers;
        final HttpHeaders newValue = updater.apply(oldValue);
        if (additionalResponseTrailersUpdater.compareAndSet(this, oldValue, newValue)) {
           return;
        }
    }
}
```

Reply:
Sure. Reduced quite a bit of code duplicates thanks to that. :-)